### DISCUSSION TITLE
Video Course On Learning Nushell

### DISCUSSION THREAD

**Discussion by trymeouteh** [nushell/nushell#13837] [Ideas]
The creation of a video course to teach you how to use Nushell.

Even if it is a paid course on Udemy.

**Comment by cboecking** [comment]
This is an interesting topic. I personally do not need a traditional training. Instead, I need a reliable way to get good answers from an AI model that references the current version I am using. 

I use aichat (https://github.com/sigoden/aichat) to assist, and I would love to strategize on ways to use its built in file + rag features to get better results. Please let me know if anyone is interested.

Tying this back to the original post, enhancing aichat (through a good model) to produce reliable answers would be a great source for a training course!

2 cents, Chuck

**Comment by ghost** [comment]
I have thought about creating a yt-video-tutorial(-series) in the past and just outline some basics best/good practices, etc.  
(and also started on text-based tutorials and  nushell-rustlings).  
however: nushell is far from stable (for example: `let-env` has changed to `$env`, `source` was deprecated for quite a while, etc). while you can just edit a section of a text-document it is a lot harder to keep a video updated.  
and the text-tutorial more or less failed since im bad at writing and most things are already in the [book](https://www.nushell.sh/book/) / [CookBook](https://www.nushell.sh/cookbook/).

Before starting anything even close to a paid course i would also want to teach some people 1:1 to learn about new-user pitfalls, etc.

<details><summary>Rambling about more problems / questions</summary>

And also question i wasn't able to answer myself when writing a script are things like the following:
* what should be part of such a turorial?
  * do i have to explain what a shell is or should i assume they already know rust and bash?
    * do i have to explain (im-)mutability?
    * do you know what a pipeline is?
    * do you know what a directory is?
    * do you know what scoping is?
  * are you interested in configuring nushell or just scripting? what about libraries?
  * do i have to explain the installation? does that include the `rustup` installation? does that include a bash-basics + windows-cmd-basics tutorial?
  * should i start with a sales-pitch for nushell?
* how long should it be? multiple parts or one video with timestamps?
* should it be a `video explains solution for last task and gives you a new task`, a reading of the book, a series of fireship-style presentations of features, Roel Van de Paar style solutions for problems, etc?
* you have to mention the differences between macos, windows, linux, etc (config-location, availability of `cat`, `aria2`, `ffmpeg`, etc)

hello world is easy, but explaining how to write tic-tac-toe or how to configure nu might already take hours.

</details>

**Reply by Lockszmith-GH** [reply]
Sorry for the long reply, started writing, and then noticed I overdid it. Hope my response is useful.

Your questions are awesome, and I believe you've highlighted a possible outline of a course with these.

> how long should it be? multiple parts or one video with timestamps?

Short videos, focusing on specific aspects (with minimal intros/outros), clear exercises/test cases and a clear learning road map (think Khan-Academy knowledge map), are probably the way human training should be done these days.

> what should be part of such a turorial? ...

The way I see it, there are at least 3 different entry points to diving into nushell

* Using it for basic functionality - the focus here would be the shell's configuration and the way basic interaction with the OS looks like.
   * Emphasis on discoverability, tab auto-completion and getting help should be the way to explain everything.
   * I would probably take the approach of looking at other shells on a POSIX system, and use that as a map: "This is how you list a directory in nushell..." etc, not a comparison (unless you feel compelled to show the differences).
   * Always send the user to find for themselves where are all the arguments and what they are by using auto-completion and help. Then demo the most common use cases. Mention if an internal command might be hiding an OS command

* The next entry point, would be the 'everything is an object' approach, where the pipeline is explained and where immutability also makes its first meaningful appearance (because it will be different, let's say, from PowerShell, where a user might be able to change the object their getting from the pipeline)

* And the last vantage point, scripting - functions, parameters, loops, logic etc.

> do i have to explain the installation? 

I wouldn't waste too much time on installation, I would choose the method that looks the same in all OSes, and point to online resources for other options, encouraging the audience to see which method suites them best.

> ...differences between macos, windows, linux...

As for OS differences, I'd mention there are, and I'll point at how to use nushell to see where those locations are in your own implementation, and possible (for advanced aka. those-who-shoot-their-own-foot-occasionally types, how to choose a location of their own), after that use examples using the variables within nushell that point to those locations.
I hate that I need to remember where the files are, when the shell already knows, and can tell me (even if it's not a short line, good reason to create a function/alias and explain that feature).

Bottom line, assume the viewer knows nothing about shell architecture, but also, assume an intelligent human with the capacity to figure things out.

> should it be a...

If we think about a university course, it's always 'mixed media' - a presentation of knowledge, which might allow for discussion, written material + other media that needs to be consumed before/after. The benefit of online content, is that the student can modify the rate of learning (even the rate of the trainer's speaking), and can skip from subject to subject more personally suitable.

When I learned PowerShell, after having experience with bash, dos/Windows cmd, c++, c#, perl, python etc... I did that with experimentation followed by reading a book.
The best part of "modern" implementation is that we have auto-complete + built in help mechanism that are way beyond what the early POSIX shells had to work with.
Teaching how to discover possible solutions, that's probably what I want to see first, and then I'd love to get guidance on structure, and best practices.

I did find that the go language online tutorials, where you get a task, you need to figure out how to solve it correctly, then you get the explanation was the best way to learn features that were 'new paradigms' for me.

Not an answer, but hope this input helps you.

> should i start with a sales-pitch for nushell?

It really depends - if it's paid content, then there isn't much use, because someone would invest in the course, only if they decided that this is what they want to learn, meaning they know what nushell is.
Free content, now that's something else, as it can be used to introduce 'outsiders', but going to the website does that, and I would say, rather efficiently.
